[NNC] Horizontally fuse all loops

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56324 [NNC] Horizontally fuse all loops**



Inlining is great if LLVM's CSE kicks in; but if a kernel has multiple outputs
(and thus multiple loops), CSE has no chance.

So, this pass "horizontally" fuses the output loops together so that CSE can go
to town. Essentially we want to turn
```
for (...) {
  output_1[] = some_complicated_expr...
}
for (...) {
  output_2[] = some_complicated_expr...
}
```

Into:
```
for (...) {
  output_1[] = complicated_expr
  output_2[] = complicated_expr. // llvm cse should take care of this
}
```

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194/)

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194)

bertmaher committed:
fuse all loops

[ghstack-poisoned]

bertmaher committed:
Update on "fuse all loops"


Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194)

[ghstack-poisoned]

bertmaher committed:
Update on "[NNC] Horizontally fuse all loops"




Inlining is great if LLVM's CSE kicks in; but if a kernel has multiple outputs
(and thus multiple loops), CSE has no chance.

So, this pass "horizontally" fuses the output loops together so that CSE can go
to town. Essentially we want to turn
```
for (...) {
  output_1[] = some_complicated_expr...
}
for (...) {
  output_2[] = some_complicated_expr...
}
```

Into:
```
for (...) {
  output_1[] = complicated_expr
  output_2[] = complicated_expr. // llvm cse should take care of this
}
```

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194/)

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194)

[ghstack-poisoned]

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1713,6 +1713,59 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+      if (!loop) {
+        // Block contains something that's not a loop.  Quit.
+        return;
+      }
```
This is going to stop if the block contains a non-loop statement anywhere. We could try to fuse loops that are together even though there are other statements before / after these loops in the block (not sure if this could happen in our IR currently, but still.. ). 

Also, if there are 4 loops and 2 of them have same bounds, we could fuse them together at least.. 

If these are not necessary for your benchmark, just leave a TODO and I will take care of them.

bertmaher replied:
Yeah this is a good idea, and might improve the benchmark even further.

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1713,6 +1713,59 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+c10::optional<std::pair<int64_t, int64_t>> loopBounds(const For* loop) {
+  auto start = IRSimplifier::simplify(loop->start());
+  auto stop = IRSimplifier::simplify(loop->stop());
+  if (!start->isConstant() || !stop->isConstant()) {
```
We should be able to fuse the loops even if the bounds are not constant, as long as they are equal. May be another refinement for later.

navahgar reviewed (approved):
LGTM. 

Suggested a couple of refinements but they could be done later.

bertmaher commented:
Hmm lots of failures.  Surprised they didn’t show up in the benchmark I was running, but looks like I’m going to have to be more careful here.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1713,6 +1713,59 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+  return true;
+}
+
+// Recursively fuse all the loops with matching bounds in `st`.  Stops fusing
+// at any level containing non-loops or non-matching bounds.  The restriction
+// on matching bounds exists to avoid inserting conditionals on the loop
+// indices where none would be needed, which would significantly complicate
+// vectorization.
+void fuseAllLoops(Stmt* st) {
```
Might be worth mentioning that in theory this pass could degrade the performance by screwing up caching of memory accesses. Benefits of CSE might overweigh this, but still.

navahgar commented:
> Hmm lots of failures. Surprised they didn’t show up in the benchmark I was running, but looks like I’m going to have to be more careful here.

I see a few of them at least resulted in a loop-carried dependence after fusing. Interesting.

navahgar commented:
> > Hmm lots of failures. Surprised they didn’t show up in the benchmark I was running, but looks like I’m going to have to be more careful here.
> 
> I see a few of them at least resulted in a loop-carried dependence after fusing. Interesting.

Looked into this.. there are couple of issues here.
1) There is a bug in the way `MemDependencyChecker` is used here. I have a fix for that.
2) The function `fuseLoops` throws an exception when it is unable to fuse. That is not the right approach, especially now that we are trying to fuse all possible loops. Specifically, it shouldn't throw exceptions when it can't fuse due a loop-carried dependence at least. We should be returning a bool flag instead to indicate whether it was fused or not.

Most of our APIs throw exceptions. @ZolotukhinM and I already discussed about this. We want to change all our APIs to return a bool flag and never throw exceptions (or throw them only on invalid IRs may be). 

I will send out a fix for both soon.

bertmaher commented:
There seems to be an additional problem on GPUs, in that one of the transforms (thread binding or flattening?) does not know what to do with incompletely fused loops.  I think it’s okay to do this transform for cpu since gpu handles horizontal fusion a different way

bertmaher committed:
Update on "[NNC] Horizontally fuse all loops"




Inlining is great if LLVM's CSE kicks in; but if a kernel has multiple outputs
(and thus multiple loops), CSE has no chance.

So, this pass "horizontally" fuses the output loops together so that CSE can go
to town. Essentially we want to turn
```
for (...) {
  output_1[] = some_complicated_expr...
}
for (...) {
  output_2[] = some_complicated_expr...
}
```

Into:
```
for (...) {
  output_1[] = complicated_expr
  output_2[] = complicated_expr. // llvm cse should take care of this
}
```

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194/)

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194)

[ghstack-poisoned]

bertmaher commented:
Fixed the GPU issue, I'm eagerly awaiting your memdep fix :-)

navahgar commented:
> Fixed the GPU issue, I'm eagerly awaiting your memdep fix :-)

Here are the 2 PRs you will need:
1. https://github.com/pytorch/pytorch/pull/56354 (fixes the memdep issue)
2. https://github.com/pytorch/pytorch/pull/56353 (changes `fuseLoops` API. You will have to update the call to this API in your PR).

I will add unit tests for these and send it our for review.

bertmaher committed:
Update on @raghavanr's fixes on "[NNC] Horizontally fuse all loops"




Inlining is great if LLVM's CSE kicks in; but if a kernel has multiple outputs
(and thus multiple loops), CSE has no chance.

So, this pass "horizontally" fuses the output loops together so that CSE can go
to town. Essentially we want to turn
```
for (...) {
  output_1[] = some_complicated_expr...
}
for (...) {
  output_2[] = some_complicated_expr...
}
```

Into:
```
for (...) {
  output_1[] = complicated_expr
  output_2[] = complicated_expr. // llvm cse should take care of this
}
```

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194/)

Differential Revision: [D27841194](https://our.internmc.facebook.com/intern/diff/D27841194)

[ghstack-poisoned]
